# Exploring Ebay Car Sales

This project should provide the opportunity to showcase a basic understanding of the pandas and numpy libraries. I will use my knowledge of these two libraries to demonstrate the following abilities:
* cleaning both text and numerical data
* indexing and manipulation of dataframe and series objects
* working understanding of many of the functions and methods available in both libraries

The dataset I will be exploring is from Dataquest.io, but it originally appeared on Kaggle. The Dataquest file is an intentionally-dirtied and randomly-sampled subset of the original file, which can be found [here](https://www.kaggle.com/orgesleka/used-cars-database/data).

In [42]:
import numpy as np
import pandas as pd
#imports the pandas and numpy libraries

In [43]:
autos = pd.read_csv('autos.csv', encoding = "Latin-1")
#reads the .csv file in as a pandas dataframe

In [44]:
autos.head()
#displays the first five rows of the autos dataframe

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [45]:
autos.info()
#displays important info about the dataframe

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

The **`.info()`** method shows us that we have 20 columns and 50000 rows, but many of our columns have null cells. Only 5 of our columns are the integer type, while the rest are currently objects. 

In [46]:
autos.columns
#displays the array of existing column names

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [47]:
copy_cols = autos.columns.copy()
#creates a copy of the columns array

In [48]:
name_dictionary = {'dateCrawled':'date_crawled', 'name':'name', 'seller':'seller', 'offerType':'offer_type', 'price':'price', 'abtest':'ab_test',
       'vehicleType': 'vehicle_type', 'yearOfRegistration': 'registration_year', 'gearbox': 'gearbox', 'powerPS':'power_ps', 'model':'model',
       'odometer':'odometer', 'monthOfRegistration': 'registration_month', 'fuelType': 'fuel_type', 'brand':'brand',
       'notRepairedDamage':'unrepaired_damage', 'dateCreated':'ad_created', 'nrOfPictures':'number_of_pictures', 'postalCode':'postal_code',
       'lastSeen':'last_seen'}
#created a dictionary of column name changes
copy_cols = copy_cols.map(name_dictionary)
#mapped dictionary onto copy of columns
autos.columns = copy_cols
#replaced columns with edited copy
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  number_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00                   0   
1         bmw              nein  2016-04-04 00:00:00                   0   
2  volkswagen              nein  2016-03-26 00:00:00                   0   
3       smart              nein  2016-03-12 00:00:00                   0   
4        ford              nein  2016-04-01 00:00:00                   0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

The first step to cleaning up the data is creating intuitive, consistent column names. Above, I renamed some of the columns to make them more descriptive and intuitive, and I ensured that snakecase was consistently applied to all column names.

In [49]:
autos.describe(include = 'all')
#returns descriptive stats for all columns

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-23 19:38:20  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  number_of_pictures   postal_code  \
count                 50000             50000.0  50000.000000   
unique                   76                 NaN           NaN   
top     2016-04-03 00:00:00                 NaN           NaN   
freq                   1946                 NaN           NaN   
mean                    NaN                 0.0  50813.627300   
std                     NaN                 0.0  25779.747957   
min                     NaN                 0.0   1067.000000   
25%                     NaN                 0.0  30451.000000   
50%                     NaN                 0.0  49577.000000   
75%                     NaN                 0.0  71540.000000   
max                     NaN                 0.0  99998.000000   

                  last_seen  
count                 50000  
unique                39481  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

Looking at the descriptive statistics, we can determine that:<br> <br>Our **`seller`** & **`offer_type`** columns both have only a couple unique entries, and the vast majority of entries (all but one) are identical. These can likely be dropped, along with **`number_of_pictures`**, which has no meaningful data due to a bug in the crawler.<br> <br>The **`date_crawled`**, **`last_seen`** and **`ad_created`** columns could be converted to a datetime type, if we end up needing them.<br> <br>For **`vehicle_type`**, **`fuel_type`**, **`brand`**, and **`model`**, we'll need to ensure consistent naming practices to make sure we're catching every instance of each type of car.<br> <br>It seems suspicious that the most common **`price`** was $0, and we'll need to convert it to a float in order to get more meaningful stats from that column.<br> <br>Also suspicious is our **`registration_year`** column, which has a minimum value of 1000,**`registration_month`**, which has a minimum value of 0, and **`postal_code`**, which has a minimum value of 1067.<br> <br>**`odometer`** will also need to be converted to a float before we can get meaningful stats out of it.

In [50]:
autos = autos.drop(columns=['seller', 'offer_type', 'number_of_pictures'])
#drops the columns without meaningful data

In [51]:
autos['date_crawled'] = pd.to_datetime(autos['date_crawled'],format= '%Y-%m-%d %H:%M:%S')
autos['last_seen'] = pd.to_datetime(autos['last_seen'],format= '%Y-%m-%d %H:%M:%S')
autos['ad_created'] = pd.to_datetime(autos['ad_created'],format= '%Y-%m-%d %H:%M:%S')
#converts the columns with timestamps to a datetime type

In [52]:
autos['vehicle_type'].unique()
#vehicle_type has consistent naming

array(['bus', 'limousine', 'kleinwagen', 'kombi', nan, 'coupe', 'suv',
       'cabrio', 'andere'], dtype=object)

In [53]:
autos['fuel_type'].unique()
#fuel_type is also consistent

array(['lpg', 'benzin', 'diesel', nan, 'cng', 'hybrid', 'elektro',
       'andere'], dtype=object)

In [54]:
autos['brand'].unique()
#brand is consistent

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'ford', 'chrysler',
       'seat', 'renault', 'mercedes_benz', 'audi', 'sonstige_autos',
       'opel', 'mazda', 'porsche', 'mini', 'toyota', 'dacia', 'nissan',
       'jeep', 'saab', 'volvo', 'mitsubishi', 'jaguar', 'fiat', 'skoda',
       'subaru', 'kia', 'citroen', 'chevrolet', 'hyundai', 'honda',
       'daewoo', 'suzuki', 'trabant', 'land_rover', 'alfa_romeo', 'lada',
       'rover', 'daihatsu', 'lancia'], dtype=object)

In [35]:
autos['model'].unique()
#model naming conventions are also consistent

array(['andere', '7er', 'golf', 'fortwo', 'focus', 'voyager', 'arosa',
       'megane', nan, 'a3', 'clio', 'vectra', 'scirocco', '3er', 'a4',
       '911', 'cooper', '5er', 'polo', 'e_klasse', '2_reihe', 'c_klasse',
       'corsa', 'mondeo', 'altea', 'a1', 'twingo', 'a_klasse', 'cl',
       '3_reihe', 's_klasse', 'sandero', 'passat', 'primera', 'fiesta',
       'wrangler', 'clubman', 'a6', 'transporter', 'astra', 'v40',
       'ibiza', 'micra', '1er', 'yaris', 'colt', '6_reihe', '5_reihe',
       'corolla', 'ka', 'tigra', 'punto', 'vito', 'cordoba', 'galaxy',
       '100', 'sharan', 'octavia', 'm_klasse', 'lupo', 'superb', 'meriva',
       'c_max', 'laguna', 'touran', '1_reihe', 'm_reihe', 'touareg',
       'seicento', 'avensis', 'vivaro', 'x_reihe', 'ducato', 'carnival',
       'boxster', 'signum', 'zafira', 'rav', 'a5', 'beetle', 'c_reihe',
       'phaeton', 'i_reihe', 'sl', 'insignia', 'up', 'civic', '80',
       'mx_reihe', 'omega', 'sorento', 'z_reihe', 'berlingo', 'clk',
       '

In [55]:
autos['price'] = autos['price'].str.replace("$","")
autos['price'] = autos['price'].str.replace(",","")
autos['price'] = autos['price'].astype(float)
#converts our price column to a float type

date_crawled                                               name  \
0 2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1 2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2 2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3 2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4 2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

    price  ab_test vehicle_type  registration_year    gearbox  power_ps  \
0  5000.0  control          bus               2004    manuell       158   
1  8500.0  control    limousine               1997  automatik       286   
2  8990.0     test    limousine               2009    manuell       102   
3  4350.0  control   kleinwagen               2007  automatik        71   
4  1350.0     test        kombi               2003    manuell         0   

    model   odometer  registration_month fuel_type       brand  \
0  andere  150,000km                   3       lpg     peugeot   
1     7er  150,000km                   6    benzin         bmw   
2    golf   70,000km                   7    benzin  volkswagen   
3  fortwo   70,000km                   6    benzin       smart   
4   focus  150,000km                   7    benzin        ford   

  unrepaired_damage ad_created  postal_code           last_seen  
0              nein 2016-03-26        79588 2016-04-06 06:45:54  
1              nein 2016-04-04        71034 2016-04-06 14:45:08  
2              nein 2016-03-26        35394 2016-04-06 20:15:37  
3              nein 2016-03-12        33729 2016-03-15 03:16:28  
4              nein 2016-04-01        39218 2016-04-01 14:38:50

In [57]:
autos['price'].describe()
#gives descriptive stats for price column

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64